flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/strategies/SubQueryInputTypeStrategyTest.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.types.inference.strategies;
+
+import org.apache.flink.table.api.DataTypes;
+import org.apache.flink.table.types.inference.InputTypeStrategiesTestBase;
+
+import java.util.stream.Stream;
+
+/** Tests for {@link SubQueryInputTypeStrategy}. */
+class SubQueryInputTypeStrategyTest extends InputTypeStrategiesTestBase {
+
+    @Override
+    protected Stream<TestSpec> testData() {
+        return Stream.of(
+                TestSpec.forStrategy("IN a set", SpecificInputTypeStrategies.IN)
```

Comment:
I'd consider the PR overall to be refactoring / moving logic around.

Given that, this test is awesome and should increase coverage.

I am not seeing tests covering (VAR)BINARY, DATETIME, CHARACTER_STRING types, or more complex types (which seem to be enabled here: https://github.com/apache/flink/pull/23652/files#diff-40e68224eeeca18aacd669a0821d31c6902c580d45bb2b9be132f804cea4a874R304-R313).

In my opinion, adding those would be a nice to have.

Reply:
It's impossible to test all possible combinations. I added some more complex cases though.

Reply:
>It's impossible to test all possible combinations. 
Ayup!  No need to try and enumerate all the possibilities.
>I added some more complex cases though.
Thanks!